ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/engine/EngineForkchoiceUpdated.java

```diff
@@ -60,101 +62,108 @@ public String getName() {
 
   @Override
   public JsonRpcResponse syncResponse(final JsonRpcRequestContext requestContext) {
+    final Object requestId = requestContext.getRequest().getId();
 
     final EngineForkchoiceUpdatedParameter forkChoice =
         requestContext.getRequiredParameter(0, EngineForkchoiceUpdatedParameter.class);
     final Optional<EnginePayloadAttributesParameter> optionalPayloadAttributes =
         requestContext.getOptionalParameter(1, EnginePayloadAttributesParameter.class);
 
+    if (mergeContext.isSyncing()) {
+      return syncingResponse(requestId);
+    }
+
+    Optional<BlockHeader> newHead =
+        protocolContext.getBlockchain().getBlockHeader(forkChoice.getHeadBlockHash());
+
+    if (newHead.isEmpty()) {
+      Optional.ofNullable(forkChoice.getHeadBlockHash())
+          .filter(hash -> !hash.equals(Hash.ZERO))
+          .ifPresent(mergeCoordinator::getOrSyncHeaderByHash);
+
+      return syncingResponse(requestId);
+    }
+
+    LOG.info(
+        "Consensus fork-choice-update: head: {}, finalized: {}, safeBlockHash: {}",
+        forkChoice.getHeadBlockHash(),
+        forkChoice.getFinalizedBlockHash(),
+        forkChoice.getSafeBlockHash());
+
     Optional<Hash> maybeFinalizedHash =
         Optional.ofNullable(forkChoice.getFinalizedBlockHash())
             .filter(finalized -> !Hash.ZERO.equals(finalized));
 
+    if (!isPartOfCanonicalChain(forkChoice.getSafeBlockHash())
+        || (maybeFinalizedHash.isPresent() && !isPartOfCanonicalChain(maybeFinalizedHash.get()))) {
+      return new JsonRpcErrorResponse(requestId, JsonRpcError.INVALID_FORKCHOICE_STATE);
+    }
+
     mergeContext.fireNewForkchoiceMessageEvent(
         forkChoice.getHeadBlockHash(), maybeFinalizedHash, forkChoice.getSafeBlockHash());
 
-    if (mergeContext.isSyncing()) {
+    // TODO: post-merge cleanup, this should be unnecessary after merge
+    if (!mergeCoordinator.latestValidAncestorDescendsFromTerminal(newHead.get())) {
       return new JsonRpcSuccessResponse(
-          requestContext.getRequest().getId(),
-          new EngineUpdateForkchoiceResult(SYNCING, null, null, Optional.empty()));
+          requestId,
+          new EngineUpdateForkchoiceResult(
+              INVALID_TERMINAL_BLOCK,
+              null,
+              null,
+              Optional.of(newHead.get() + " did not descend from terminal block")));
     }
 
-    LOG.info(
-        "Consensus fork-choice-update: head: {}, finalized: {}",
-        forkChoice.getHeadBlockHash(),
-        forkChoice.getFinalizedBlockHash());
+    ForkchoiceResult result =
+        mergeCoordinator.updateForkChoice(
+            forkChoice.getHeadBlockHash(), forkChoice.getFinalizedBlockHash());
```

Comment:
we should add safeBlockHash to the mergeContext while we are here.  We will need it for json-rpc eth_getBlockByNumber

Reply:
We save it now in mergeContext and in the blockchain storage